src/socket.rs

```diff
@@ -88,18 +81,18 @@ impl NlSocket {
 
     /// Serialize and send Rust `NlMsg` type
     pub fn sendmsg<I: Nl, T: Nl>(&mut self, mut msg: NlHdr<I, T>, flags: i32)
-                                        -> Result<isize, NlError> {
+                                        -> Result<usize, NlError> {
         let mut state = NlSerState::new();
         try!(msg.serialize(&mut state));
-        let len = try!(self.send(state.into_inner().as_slice(), flags));
-        Ok(len)
+        try!(self.send(state.into_inner().as_slice(), flags))
     }
 
     /// Receive and deserialize Rust `NlMsg` type
-    pub fn recvmsg<I: Nl, T: Nl>(&mut self, len: Option<usize>, flags: i32)
+    pub fn recvmsg<I: Nl, T: Nl>(&mut self, len: usize, flags: i32)
                                        -> Result<NlHdr<I, T>, NlError> {
-        let mut buf = try!(self.recv(len, flags));
-        let msg = try!(<NlHdr<I, T> as Nl>::deserialize(&mut NlDeState::new(buf.as_mut_slice())));
+        let mut buf = vec![0; len];
```

Comment:
Here is some internal allocation left. I didn't want to change too much before getting some feedback from you what you think of this type of change.

Reply:
For such a convenience function it might be fully OK to do internal allocation. As long as there is a way to use the library without it. That is separate, easy to use, `recv` and `deserialize`.

Problem here is that `buf` is dropped after each call. So it's quite expensive in the current state. Having to do a heap allocation for each read packet.